File: src/app/workspace.c
#include "sounds/workspace.h"

static const SoundWorkspace workspaces[] = {
    SOUND_WORKSPACE_LIVE,
    SOUND_WORKSPACE_CLIPS,
    SOUND_WORKSPACE_SPECTRUM,
    SOUND_WORKSPACE_BAND,
    SOUND_WORKSPACE_COMPARE,
};

int sound_workspace_count(void) {
    return (int)(sizeof(workspaces) / sizeof(workspaces[0]));
}

SoundWorkspace sound_workspace_at(int index) {
    if (index < 0 || index >= sound_workspace_count()) {
        return SOUND_WORKSPACE_LIVE;
    }

    return workspaces[index];
}

int sound_workspace_index(SoundWorkspace workspace) {
    for (int i = 0; i < sound_workspace_count(); ++i) {
        if (workspaces[i] == workspace) {
            return i;
        }
    }

    return 0;
}

SoundWorkspace sound_workspace_offset(SoundWorkspace workspace, int offset) {
    int count = sound_workspace_count();
    int index = sound_workspace_index(workspace) + offset;

    while (index < 0) {
        index += count;
    }

    return sound_workspace_at(index % count);
}

const char *sound_workspace_name(SoundWorkspace workspace) {
    switch (workspace) {
        case SOUND_WORKSPACE_LIVE:
            return "LIVE SPECTROGRAM";
        case SOUND_WORKSPACE_CLIPS:
            return "CLIPS";
        case SOUND_WORKSPACE_SPECTRUM:
            return "WHOLE SPECTRUM";
        case SOUND_WORKSPACE_BAND:
            return "BAND LAB";
        case SOUND_WORKSPACE_COMPARE:
            return "COMPARE";
        case SOUND_WORKSPACE_COUNT:
            break;
    }

    return "LIVE SPECTROGRAM";
}

const char *sound_workspace_short_name(SoundWorkspace workspace) {
    switch (workspace) {
        case SOUND_WORKSPACE_LIVE:
            return "LIVE";
        case SOUND_WORKSPACE_CLIPS:
            return "CLIPS";
        case SOUND_WORKSPACE_SPECTRUM:
            return "SPECTRUM";
        case SOUND_WORKSPACE_BAND:
            return "BAND";
        case SOUND_WORKSPACE_COMPARE:
            return "COMPARE";
        case SOUND_WORKSPACE_COUNT:
            break;
    }

    return "LIVE";
}

File: src/app/workspace.zig
//! Workspace/tab state.
//! Rewrite target; C reference: src_c/src/app/workspace.c

pub const Workspace = enum {
    live,
    recordings,
    trim,
    spectrum,
    band_lab,

    pub fn index(self: Workspace) usize {
        return @intFromEnum(self);
    }

    pub fn name(self: Workspace) []const u8 {
        return switch (self) {
            .live => "LIVE SPECTROGRAM",
            .recordings => "RECORDINGS",
            .trim => "TRIM",
            .spectrum => "WHOLE SPECTRUM",
            .band_lab => "BAND LAB",
        };
    }

    pub fn shortName(self: Workspace) []const u8 {
        return switch (self) {
            .live => "LIVE",
            .recordings => "RECS",
            .trim => "TRIM",
            .spectrum => "SPECTRUM",
            .band_lab => "BAND",
        };
    }

    pub fn offset(self: Workspace, delta: isize) Workspace {
        return workspaceOffset(self, delta);
    }
};

pub const order = [_]Workspace{
    .live,
    .recordings,
    .trim,
    .spectrum,
    .band_lab,
};

pub const count = order.len;

pub fn at(index: isize) Workspace {
    if (index < 0) return .live;
    const i: usize = @intCast(index);
    if (i >= order.len) return .live;
    return order[i];
}

pub fn workspaceOffset(workspace: Workspace, delta: isize) Workspace {
    const workspace_count: isize = @intCast(order.len);
    var index: isize = @as(isize, @intCast(workspace.index())) + delta;
    while (index < 0) index += workspace_count;
    return order[@intCast(@mod(index, workspace_count))];
}

test "workspace order and wrapping match the C API" {
    const std = @import("std");

    try std.testing.expectEqual(@as(usize, 5), count);
    try std.testing.expectEqual(Workspace.live, at(-1));
    try std.testing.expectEqual(Workspace.live, at(99));
    try std.testing.expectEqual(Workspace.recordings, Workspace.live.offset(1));
    try std.testing.expectEqual(Workspace.live, Workspace.band_lab.offset(1));
    try std.testing.expectEqual(Workspace.band_lab, Workspace.live.offset(-1));
    try std.testing.expectEqualStrings("RECORDINGS", Workspace.recordings.name());
    try std.testing.expectEqualStrings("BAND", Workspace.band_lab.shortName());
}
